src/H5Dchunk.c

```diff
@@ -5603,156 +5607,73 @@ H5D__chunk_collective_fill(const H5D_t *dset, H5D_chunk_coll_fill_info_t *chunk_
-            partial_fill_buf_disp = MPI_Aint_diff(partial_fill_buf_addr, fill_buf_addr);
-#else
-            partial_fill_buf_disp = partial_fill_buf_addr - fill_buf_addr;
-#endif
-
-            /*
-             * Allocate all-zero block displacements array. If a block's displacement
-             * is left as zero, that block will be written to from the regular fill
-             * buffer. If a block represents an unfiltered partial edge chunk, its
-             * displacement will be set so that the block is written to from the
-             * unfiltered fill buffer.
-             */
-            if (NULL == (block_disps = (MPI_Aint *)H5MM_calloc((size_t)(blocks + 1) * sizeof(MPI_Aint))))
-                HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate block displacements buffer");
-        }
 
-        /*
-         * Perform initial scan of chunk info list to:
-         *  - make sure that chunk addresses are monotonically non-decreasing
-         *  - check if all blocks have the same length
-         */
-        for (i = 1; i < chunk_fill_info->num_chunks; i++) {
-            if (chunk_fill_info->chunk_info[i].addr < chunk_fill_info->chunk_info[i - 1].addr)
-                need_sort = true;
+        if (NULL == (io_addrs = H5MM_malloc((size_t)(blocks + 1) * sizeof(*io_addrs))))
+            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                        "couldn't allocate space for I/O addresses vector");
 
-            if (chunk_fill_info->chunk_info[i].chunk_size != chunk_fill_info->chunk_info[i - 1].chunk_size)
-                all_same_block_len = false;
-        }
+        if (NULL == (io_wbufs = H5MM_malloc((size_t)(blocks + 1) * sizeof(*io_wbufs))))
+            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "couldn't allocate space for I/O buffers vector");
+    }
 
-        if (need_sort)
-            qsort(chunk_fill_info->chunk_info, chunk_fill_info->num_chunks,
-                  sizeof(struct chunk_coll_fill_info), H5D__chunk_cmp_coll_fill_info);
+    /*
+     * Perform initial scan of chunk info list to:
+     *  - make sure that chunk addresses are monotonically non-decreasing
+     *  - check if all blocks have the same length
+     */
+    for (i = 1; i < chunk_fill_info->num_chunks; i++) {
+        if (chunk_fill_info->chunk_info[i].addr < chunk_fill_info->chunk_info[i - 1].addr)
+            need_sort = true;
 
-        /* Allocate buffer for block lengths if necessary */
-        if (!all_same_block_len)
-            if (NULL == (block_lens = (int *)H5MM_malloc((size_t)(blocks + 1) * sizeof(int))))
-                HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate chunk lengths buffer");
+        if (chunk_fill_info->chunk_info[i].chunk_size != chunk_fill_info->chunk_info[i - 1].chunk_size)
+            all_same_block_len = false;
+    }
 
-        for (i = 0; i < (size_t)blocks; i++) {
-            size_t idx = i + (size_t)(mpi_rank * blocks);
+    if (need_sort)
```

Comment:
Could you add some more comments to the blocks that don't have them, explaining what the code is doing?

Reply:
Also, for this block in particular, add something like this note (I just discussed this with Scot):

Note that we sort all of the chunks here, and not just a subset corresponding to this rank. We do this since we have found MPI I/O to work better when each rank writes blocks that are contiguous in the file, and by sorting the full list we maximize the chance of that happening.

Reply:
It seems like we could potentially be making extra work for ourselves now by sorting here since the MPI I/O VFD is already scanning the chunk list to check for this anyway and we don't necessarily need to impose the monotonically non-decreasing file displacement requirement on other parallel VFDs. That said, since the sort has to happen at some point anyway if the addresses are out of order, I'm not sure if the above note is really needed here.